pkg/querier/queryrange/splitters.go

```diff
@@ -0,0 +1,269 @@
+	if err != nil {
+		return 0, err
+	}
+	if maxRange > interval {
+		return maxRange, nil
+	}
+	return interval, nil
+}
+
+// Round up to the step before the next interval boundary.
+func (s *metricQuerySplitter) nextIntervalBoundary(t time.Time, step int64, interval time.Duration) time.Time {
+	stepNs := step * 1e6
+	nsPerInterval := interval.Nanoseconds()
+	startOfNextInterval := ((t.UnixNano() / nsPerInterval) + 1) * nsPerInterval
+	// ensure that target is a multiple of steps away from the start time
+	target := startOfNextInterval - ((startOfNextInterval - t.UnixNano()) % stepNs)
+	if target == startOfNextInterval {
+		target -= stepNs
+	}
+	return time.Unix(0, target)
+}
+
+func (s *metricQuerySplitter) split(execTime time.Time, tenantIDs []string, r queryrangebase.Request, interval time.Duration) ([]queryrangebase.Request, error) {
+	var reqs []queryrangebase.Request
+
+	lokiReq := r.(*LokiRequest)
+
+	interval, err := s.reduceSplitIntervalForRangeVector(lokiReq, interval)
+	if err != nil {
+		return nil, err
+	}
+
+	// step align start and end time of the query. Start time is rounded down and end time is rounded up.
+	stepNs := r.GetStep() * 1e6
+	startNs := lokiReq.StartTs.UnixNano()
+	start := time.Unix(0, startNs-startNs%stepNs)
+
+	endNs := lokiReq.EndTs.UnixNano()
+	if mod := endNs % stepNs; mod != 0 {
+		endNs += stepNs - mod
+	}
+	end := time.Unix(0, endNs)
+
+	lokiReq = lokiReq.WithStartEnd(start, end).(*LokiRequest)
+
+	factory := func(start, end time.Time) {
+		reqs = append(reqs, &LokiRequest{
+			Query:     lokiReq.Query,
+			Limit:     lokiReq.Limit,
+			Step:      lokiReq.Step,
+			Interval:  lokiReq.Interval,
+			Direction: lokiReq.Direction,
+			Path:      lokiReq.Path,
+			StartTs:   start,
+			EndTs:     end,
+			Plan:      lokiReq.Plan,
+		})
+	}
+
+	// step is >= configured split interval, let us just split the query interval by step
+	if lokiReq.Step >= interval.Milliseconds() {
+		util.ForInterval(time.Duration(lokiReq.Step*1e6), lokiReq.StartTs, lokiReq.EndTs, false, factory)
+
+		return reqs, nil
+	}
+
+	// Treat range of given query within the `query_ingesters_within` window differently by splitting using the `split_ingester_queries_by_interval`
+	// instead of the default `split_queries_by_interval`; rebound the start/end time after doing so to build intervals
+	// for queries outside the `query_ingesters_within` window.
+	//
+	// The given factory is responsible for building the splits and appending to reqs.
+	newStart, newEnd := buildIngesterQuerySplitsAndRebound(execTime, s.limits, s.iqo, tenantIDs, lokiReq, factory, false)
+	if newStart.Equal(newEnd) {
+		// Nothing to do, there are no more intervals to process.
+		return reqs, nil
+	}
+
+	// copy the splits, reset the results
+	ingesterSplits := reqs
+	reqs = nil
+
+	for start := newStart; start.Before(newEnd); start = s.nextIntervalBoundary(start, r.GetStep(), interval).Add(time.Duration(r.GetStep()) * util.SplitGap) {
```

Comment:
why not leave this as `time.Millisecond` since we are essentially converting it to ns?
I feel `SplitGap` is not the right thing to use here